odoo/addons/base/models/ir_async.py

```diff
@@ -0,0 +1,299 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+import itertools
+import json
+import functools
+import logging
+import re
+import traceback
+import threading
+import odoo
+import odoo.exceptions
+from odoo import models, fields, api
+from odoo.http import serialize_exception
+
+_logger = logging.getLogger(__name__)
+
+
+BASE_VERSION = odoo.modules.load_information_from_description_file('base')['version']
+class BadVersion(Exception): pass
+class BadModuleState(Exception): pass
+
+S_CREATED = "created"        # The task has been enqueued
+S_PROCESSING = "processing"  # The worker begins to process the task
+S_READY = "ready"            # The task completed with an action for the user
+S_COMPLETED = "completed"    # The user executed the action
+S_FAILED = "failed"          # The task failed with an error
+STATUS = [S_CREATED, S_PROCESSING, S_READY, S_COMPLETED, S_FAILED]
+
+
+class ir_async(models.Model):
+    _name = 'ir.async'
+    _description = 'Asynchrone jobs'
+
+    name = fields.Char(default="Unamed background task")
+    res_model = fields.Char()
+    res_ids = fields.Text()
+    method = fields.Char()
+    args = fields.Text()
+    kwargs = fields.Text()
+    env_uid = fields.Many2one('res.users', string='Scheduler User')
+    env_context = fields.Text()
+    traceback = fields.Text()
+    notify = fields.Boolean()
+
+    def call(self, bound_method, *args, **kwargs):
+        """
+        Partial-like API to create asynchronous jobs for model methods.
+
+        Ensure the method is called in an independant worker using a
+        copy of the current environnement (user, context, recordset).
+
+        The method return value is discarded unless the user set the
+        ``async_notify`` context key. When the context key is set, the
+        user is notified on the bus twice: (i) when the job starts and
+        (ii) when it completes. The message sent on completion is a JSON
+        payload just like normal RPC responses.
+
+        The ``async`` javascript service can be used to correctly
+        resolve asynchronous jobs on the client-side.
+        """
+        recs = getattr(bound_method, '__self__', None)
+        if recs is None or not isinstance(recs, models.BaseModel):
+            raise TypeError("You can only create an async task on a recordset")
+        model = recs.__class__
+
+        job = self.sudo().create({
+            'name': self.env.context.get('async_notify', False),
+            'res_model': model._name,
+            'res_ids': json.dumps(recs._ids),
+            'method': bound_method.__name__,
+            'args': json.dumps(args),
+            'kwargs': json.dumps(kwargs),
+            'env_uid': int(recs.env.uid),
+            'env_context': json.dumps(recs.env.context),
+            'traceback': self._merge_traceback(traceback.format_stack()[:-1]),
+            'notify': self.env.context.get('async_notify', False),
+        })
+        job.flush()
```

Comment:
Really not fond of this shortcut. Or having to to through through the context to pass a glorified parameter through the call (even less so as it's apparently used as both a name and a boolean flag? is `name` really supposed to be `context['async_notify']`? That seems really strange).

Reply:
I don't like it either... let's do like Thread() and Process(), ditch *args and **kwargs for decicated args and kwargs normal arguments